src/app/needs/next-actions/components/NextActionsDisplay.tsx

```diff
@@ -0,0 +1,212 @@
+  const router = useRouter();
+  const [highlightedNeeds, setHighlightedNeeds] = useState<NeedDocument[]>([]);
+  const [relatedNextActions, setRelatedNextActions] = useState<NextActionDocument[]>([]);
+  const [chainEnd, setChainEnd] = useState(0); 
+  // chainEnd increments after toggling actions to refetch and update UI
+
+  useEffect(() => {
+    async function fetchData() {
+      // 1. Fetch all needs (as RxDocuments), then to JSON
+      const needsDocs = await database.getFromDb("needs");
+      const allNeeds = needsDocs.map(doc => doc.toJSON() as NeedDocument);
+
+      // 2. Highlighted needs check: selectedExpiry in the future means highlighted
+      const now = new Date();
+      const filteredNeeds = allNeeds.filter(need => {
+        const expiry = new Date(need.selectedExpiry);
+        return expiry > now && need.priority && need.priority.order > 0;
+      });
+
+      setHighlightedNeeds(filteredNeeds);
+
+      // 3. Fetch all next actions (as RxDocuments), then to JSON
+      const nextActionsDocs = await database.getFromDb("next_actions");
+      const allNextActions = nextActionsDocs.map(doc => doc.toJSON() as NextActionDocument);
+
+      // 4. Filter next actions to only those related to highlighted needs
+      const highlightedNeedIds = new Set(filteredNeeds.map(need => need.id));
+      const filteredNextActions = allNextActions.filter(na => highlightedNeedIds.has(na.need));
+
+      setRelatedNextActions(filteredNextActions);
+    }
+
+    fetchData();
+  }, [database, chainEnd]);
+
+  // Group and sort highlighted needs by priority
+  const priorityGroups = useMemo(() => {
+    if (highlightedNeeds.length === 0) return [];
+
+    const needsWithPriority = highlightedNeeds.filter(
+      need => need.priority && need.priority.order
+    );
+
+    const sortedNeeds = needsWithPriority.sort((a, b) =>
+      a.priority!.order - b.priority!.order
+    );
+
+    const groupsMap: Record<number, { priority: { order: number; name: string}; needs: NeedDocument[] }> = {};
+    for (const need of sortedNeeds) {
+      const order = need.priority!.order;
+      if (!groupsMap[order]) {
+        groupsMap[order] = {
+          priority: need.priority!,
+          needs: []
+        };
+      }
+      groupsMap[order].needs.push(need);
+    }
+
+    return Object.values(groupsMap);
+  }, [highlightedNeeds]);
+
+  // Helper to get next actions for a given need
+  const getActionsForNeed = (needId: string) => {
+    return relatedNextActions.filter(action => action.need === needId);
+  };
+
+  // onToggleAction logic
+  const onToggleAction = async (action: NextActionDocument) => {
+    const highlighted = new Date(action.selectedExpiry) > new Date(action.timestamp);
+    const collectionName = "next_actions";
+
+    if (highlighted) {
+      // Un-highlight action:
+      // Remove last selectedTimestamp
+      const updatedTimestamps = [...action.selectedTimestamps];
+      updatedTimestamps.pop();
```

Comment:
Why are you using pop here? Not sure we should be reducing this array, we should just keep appending the date now when the user clicks on it to select it. If a user deselects we just update the selectedExpiry to be now or in the past (as you have done below using the document timestamp), but this does not change the fact that the user did select this next action previously and we want to keep that data if we are to use it as insight

Reply:
Disagree, because it doesn't handle accidental selection. Also deselecting indicates no longer intending to do the action (and I don't want to track changed intention) which is different to deselecting a need (which indicates it's now met and should be tracked)